server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java

```diff
@@ -0,0 +1,362 @@
+                    publication.onFaultyNode(n2);
+                }
+                if (remainingActions.decrementAndGet() == injectFaultAt) {
+                    publication.onFaultyNode(n2);
+                }
+                if (e.getKey().equals(n2) == false || randomBoolean()) {
+                    nodeResolver.apply(e.getKey()).coordinationState.handleCommit(publication.applyCommit);
+                    e.getValue().onResponse(TransportResponse.Empty.INSTANCE);
+                }
+            }
+        });
+
+        // we need to complete publication by failing the node
+        if (failNodeWhenCommitting && publicationDidNotMakeItToNode2 && remainingActions.get() > injectFaultAt) {
+            publication.onFaultyNode(n2);
+        }
+
+        assertTrue(publication.completed);
+        assertTrue(publication.committed);
+
+        publication.onFaultyNode(randomFrom(n1, n3)); // has no influence
+
+        List<Tuple<DiscoveryNode, Throwable>> errors = ackListener.awaitErrors(0L, TimeUnit.SECONDS);
+        assertThat(errors.size(), equalTo(1));
+        assertThat(errors.get(0).v1(), equalTo(n2));
+    }
+
+    public void testClusterStatePublishingFailsOrTimesOutBeforeCommit() throws InterruptedException {
+        VotingConfiguration config = new VotingConfiguration(Sets.newHashSet(n1.getId(), n2.getId()));
+        initializeCluster(config);
+
+        AssertingAckListener ackListener = new AssertingAckListener(nodes.size());
+        DiscoveryNodes discoveryNodes = DiscoveryNodes.builder().add(n1).add(n2).add(n3).localNodeId(n1.getId()).build();
+        MockPublication publication = node1.publish(CoordinationStateTests.clusterState(1L, 2L,
+            discoveryNodes, config, config, 42L), ackListener, Collections.emptySet());
+
+        publication.pendingPublications.entrySet().stream().collect(shuffle()).forEach(e -> {
+            if (e.getKey().equals(n2)) {
+                if (randomBoolean()) {
+                    publication.onTimeout();
+                } else {
+                    e.getValue().onFailure(new TransportException(new Exception("dummy failure")));
+                }
+                assertTrue(publication.completed);
+                assertFalse(publication.committed);
+            } else if (randomBoolean()) {
+                PublishResponse publishResponse = nodeResolver.apply(e.getKey()).coordinationState.handlePublishRequest(
+                    publication.publishRequest);
+                e.getValue().onResponse(new PublishWithJoinResponse(publishResponse, Optional.empty()));
+            }
+        });
+
+        assertThat(publication.pendingCommits.keySet(), equalTo(Collections.emptySet()));
+        assertNull(publication.applyCommit);
+        assertTrue(publication.completed);
+        assertFalse(publication.committed);
+
+        List<Tuple<DiscoveryNode, Throwable>> errors = ackListener.awaitErrors(0L, TimeUnit.SECONDS);
+        assertThat(errors.size(), equalTo(3));
```

Comment:
Can we assert something about the errors themselves, and that there were no successful acks? (also throughout the other tests)

Reply:
sure, gave this a try in b711edb4894